src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-                            if (complete) {
-                                parentSubscriber.completeInner(this);
-                            } else {
-                                try {
-                                    parentSubscriber.actual.onNext(t);
-                                } catch (Throwable e) {
-                                    // special error handling due to complexity of merge
-                                    onError(OnErrorThrowable.addValueAsLastCause(e, t));
-                                }
-                                emitted++;
-                                MergeProducer.REQUESTED.decrementAndGet(producer);
-                            }
-                        } else {
-                            // no requests available, so enqueue it
-                            enqueue = true;
-                        }
+                    queue.onNext(parent.nl.next(t));
+                } catch (MissingBackpressureException ex) {
+                    unsubscribe();
+                    onError(ex);
+                    return;
+                } catch (IllegalStateException ex) {
+                    if (!isUnsubscribed()) {
+                        unsubscribe();
+                        onError(ex);
                     }
-                } finally {
-                    drain = parentSubscriber.releaseEmitLock();
-                }
-                // request upstream what we just emitted
-                if(emitted > 0) {
-                    request(emitted);
+                    return;
                 }
+                parent.emit();
             }
-            if (enqueue) {
-                enqueue(t, complete);
-                drain = true;
+            @Override
+            public void onError(Throwable e) {
+                done = true;
+                parent.onError(e);
+            }
+            @Override
+            public void onCompleted() {
+                done = true;
+                parent.emit();
             }
-            if (drain) {
-                /**
-                 * This extra check for whether to call drain is ugly, but it helps:
-                 * <pre> {@code
-                 * Without:
-                 * r.o.OperatorMergePerf.mergeNSyncStreamsOfN     1000  thrpt         5       61.812        1.455    ops/s
-                 * 
-                 * With:
-                 * r.o.OperatorMergePerf.mergeNSyncStreamsOfN     1000  thrpt         5       78.795        1.766    ops/s
-                 * } </pre>
-                 */
-                parentSubscriber.drainQueuesIfNeeded();
+            
+            void requestMore(long n) {
+                request(n);
             }
         }
-
-        private void enqueue(T t, boolean complete) {
+    }
+    /**
+     * Subscriber that requests Long.MAX_VALUE from upstream and doesn't queue. 
+     * @param <T>
+     */
+    static final class FastInnerSubscriber<T> extends Subscriber<T> {
```

Comment:
The idea of the unbounded mode is to bypass queueing as much as possible. Naturally, FastInnerSubscriber can request in batches so the serialization buffer doesn't grow unbounded, but what should be the default batch size?